src/foam/nanos/auth/SubjectAuthService.js

```diff
@@ -0,0 +1,55 @@
+/**
+ * @license
+ * Copyright 2023 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.auth',
+  name: 'SubjectAuthService',
+  extends: 'foam.nanos.auth.ProxyAuthService',
+
+  documentation: 'A decorator to check auth against every user in subject user path',
+
+  javaImports: [
+    'foam.nanos.auth.Subject',
+    'foam.nanos.auth.User',
+    'foam.util.Auth',
+    'java.util.List',
+    'java.util.HashMap',
+    'java.util.Map'
+  ],
+
+  methods: [
+    {
+      name: 'check',
+      javaCode: `
+        List<User> userPath = ((Subject) x.get("subject")).getUserPath();
+        if ( userPath.size() == 0 ) return false;
+        
+        if ( getDelegate().check(x, permission) ) return true;
+
+        if ( userPath.size() == 1 ) return false;
+
+        if ( userPath.size() == 2 ) {
+          if ( userPath.get(0).getId() == userPath.get(1).getId() ) {
```

Comment:
This can't happen. It won't push the same id onto the path twice.

Instead do:
if ( userPath.size() == 2 ) {
  return getDelegate().checkUser(x, userPath.get(0), permission);
}

This avoids creating the seen HashMap.
}

Reply:
yeah thx